packages/fxa-auth-db-mysql/db-server/test/backend/db_tests.js

```diff
@@ -4470,6 +4470,100 @@ module.exports = function(config, DB) {
         }
       });
 
+      it('should reactivate subscriptions', async () => {
+        await db.createAccountSubscription(
+          account.uid,
+          subscriptionIds[22],
+          'prod0',
+          Date.now()
+        );
+        await db.cancelAccountSubscription(
+          account.uid,
+          subscriptionIds[22],
+          Date.now()
+        );
+        await db.reactivateAccountSubscription(account.uid, subscriptionIds[22]);
+
+        const subscriptions = await db.fetchAccountSubscriptions(account.uid);
+
+        assert.lengthOf(subscriptions, 1);
+        assert.deepEqual(
+          pickSet(subscriptions, 'subscriptionId'),
+          new Set([subscriptionIds[22]])
+        );
+        assert.deepEqual(
+          pickSet(subscriptions, 'cancelledAt'),
+          new Set([null])
+        );
+      });
+
+      it('should fail to reactivate a non-existent subscription', async () => {
+        let failed = false;
+        try {
+          await db.reactivateAccountSubscription(
+            account.uid,
+            subscriptionIds[23],
+            Date.now()
+          );
+        } catch (err) {
+          failed = true;
+          assert.equal(err.errno, 116);
+        }
+        assert(failed);
+      });
+
+      it('should fail to reactivate a non-cancelled subscription', async () => {
+        await db.createAccountSubscription(
+          account.uid,
+          subscriptionIds[24],
+          'prod0',
+          Date.now()
+        );
+
+        let failed = false;
+        try {
+          await db.reactivateAccountSubscription(
+            account.uid,
+            subscriptionIds[24],
+            Date.now()
+          );
+        } catch (err) {
+          failed = true;
+          assert.equal(err.errno, 116);
+        }
+        assert(failed);
+      });
+
+      it('should fail to reactivate a reactivated subscription', async () => {
+        await db.createAccountSubscription(
+          account.uid,
+          subscriptionIds[25],
+          'prod0',
+          Date.now()
+        );
+        await db.cancelAccountSubscription(
+          account.uid,
+          subscriptionIds[25],
+          Date.now()
+        );
+        await db.reactivateAccountSubscription(
+          account.uid,
+          subscriptionIds[25],
+        );
+
+        let failed = false;
+        try {
+          await db.reactivateAccountSubscription(
+            account.uid,
+            subscriptionIds[25],
+          );
```

Comment:
I've seen other tests where we `assert.fail()` in a try if we expect it to error. This is the first time I've seen this pattern of a failed bool that is asserted later. Is there a preference in the code-base that we should aim for?

Reply:
I disagree with the use of `assert.fail()` inside `try` blocks because the assert error gets caught by the `catch`, so it relies on secondary assertions there to actually fail the test. That makes it harder to diagnose the true causes of failure and it's even conceivable that a failure might accidentally be swallowed if assertions inside the `catch` don't differentiate it from their expected error type.

@lmorchard previously made the same point to me in https://github.com/mozilla/fxa/pull/1112#discussion_r284338002 and I changed my approach in that PR. But my opinion has hardened since then and I now think using `assert.fail()` like that is objectively wrong.

Reply:
Sounds good to me. Can we codify this as a best practice in a code style doc? I'm not sure an ADR makes sense, but ideally things like this should be in our own code style document.

Reply:
Also, is there a reason we can't use `https://www.chaijs.com/plugins/chai-as-promised/` with `assert.isRejected` or `isFulfilled`?

Reply:
Oh, probably not, other than lack of awareness. Let's add it next opportunity.